src/deluge/gui/views/performance_session_view.cpp

```diff
@@ -0,0 +1,1931 @@
+		//<resetValue>
+		storageManager.writeTag(PERFORM_DEFAULTS_HOLD_RESETVALUE_TAG, kNoSelection);
+
+		backupXMLDefaultFXPress[xDisplay].padPressHeld = false;
+		backupXMLDefaultFXPress[xDisplay].yDisplay = kNoSelection;
+		backupXMLDefaultFXPress[xDisplay].previousKnobPosition = kNoSelection;
+		backupXMLDefaultFXPress[xDisplay].currentKnobPosition = kNoSelection;
+	}
+
+	storageManager.writeClosingTag(PERFORM_DEFAULTS_HOLD_TAG);
+}
+
+/// backup current layout, load saved layout, log layout change, update change status
+void PerformanceSessionView::loadPerformanceViewLayout() {
+	char modelStackMemory[MODEL_STACK_MAX_SIZE];
+	ModelStackWithThreeMainThings* modelStack = currentSong->setupModelStackWithSongAsTimelineCounter(modelStackMemory);
+
+	backupPerformanceLayout();
+	resetPerformanceView(modelStack);
+	if (successfullyReadDefaultsFromFile) {
+		readDefaultsFromBackedUpFile();
+	}
+	else {
+		readDefaultsFromFile();
+	}
+	logPerformanceLayoutChange();
+	updateLayoutChangeStatus();
+	uiNeedsRendering(this);
+}
+
+/// re-read defaults from backed up XML in memory in order to reduce SD Card IO
+void PerformanceSessionView::readDefaultsFromBackedUpFile() {
+	for (int32_t xDisplay = 0; xDisplay < kDisplayWidth; xDisplay++) {
+		memcpy(&layoutForPerformance[xDisplay], &backupXMLDefaultLayoutForPerformance[xDisplay],
+		       sizeParamsForPerformance);
+
+		memcpy(&fxPress[xDisplay], &backupXMLDefaultFXPress[xDisplay], sizeFXPress);
+
+		for (int32_t yDisplay = 0; yDisplay < kDisplayHeight; yDisplay++) {
+			defaultFXValues[xDisplay][yDisplay] = backupXMLDefaultFXValues[xDisplay][yDisplay];
+		}
+
+		initializeHeldFX(xDisplay);
+	}
+}
+
+/// read defaults from XML
+void PerformanceSessionView::readDefaultsFromFile() {
+	//no need to keep reading from SD card after first load
+	if (successfullyReadDefaultsFromFile) {
+		return;
+	}
+
+	FilePointer fp;
+	//PerformanceView.XML
+	bool success = storageManager.fileExists(PERFORM_DEFAULTS_XML, &fp);
+	if (!success) {
+		loadDefaultLayout();
+		return;
+	}
+
+	//<defaults>
+	int32_t error = storageManager.openXMLFile(&fp, PERFORM_DEFAULTS_TAG);
+	if (error) {
+		loadDefaultLayout();
+		return;
+	}
+
+	char const* tagName;
+	//step into the <defaultFXValues> tag
+	while (*(tagName = storageManager.readNextTagOrAttributeName())) {
+		if (!strcmp(tagName, PERFORM_DEFAULTS_FXVALUES_TAG)) {
+			readDefaultFXValuesFromFile();
+		}
+		storageManager.exitTag();
+	}
+
+	storageManager.closeFile();
+
+	if (!successfullyReadDefaultsFromFile) {
+		backupPerformanceLayout();
+		logPerformanceLayoutChange();
+	}
+
+	successfullyReadDefaultsFromFile = true;
+}
+
+/// if no XML file exists, load default layout (paramKind, paramID, xDisplay, yDisplay, rowColour, rowTailColour)
+void PerformanceSessionView::loadDefaultLayout() {
+	for (int32_t xDisplay = 0; xDisplay < kDisplayWidth; xDisplay++) {
+		memcpy(&layoutForPerformance[xDisplay], &defaultLayoutForPerformance[xDisplay], sizeParamsForPerformance);
```

Comment:
You should implement a copy method instead of using memcpy - memcpy is brittle and any speed differences won't matter here

Reply:
you mean like this?

`copyLayout(layoutForPerformance[xDisplay], defaultLayoutForPerformance[xdisplay]);`

```
PerformanceSessionView::copyLayout(ParamForPerformance& layoutTo, ParamForPerformance& layoutFrom) {

layoutTo.paramKind = layoutFrom.paramKind;
layoutTo.paramID = layoutFrom.paramID;
layoutTo.xDisplay = layoutFrom.xDisplay;
layoutTo.yDisplay = layoutFrom.yDisplay;
layoutTo.rowColour[0] = layoutFrom.rowColour[0];
layoutTo.rowColour[1] = layoutFrom.rowColour[1];
layoutTo.rowColour[2] = layoutFrom.rowColour[2];
layoutTo.rowTailColour[0] = layoutFrom.rowTailColour[0];
layoutTo.rowTailColour[1] = layoutFrom.rowTailColour[1];
layoutTo.rowTailColour[2] = layoutFrom.rowTailColour[2];

}
```

Reply:
I need to look in more detail but this would generate a lot more code compared to memcpy and is more error prone, any particular reason?

Reply:
> I need to look in more detail but this would generate a lot more code compared to memcpy and is more error prone, any particular reason?

Yea cause I'd need to create copy functions for each type of struct I'm trying to copy (so 3 copy functions). I initially thought of this but thought memcpy was cleaner so I went with memcpy.